java-spiffe-core/src/main/java/spiffe/bundle/x509bundle/X509Bundle.java

```diff
@@ -0,0 +1,135 @@
+        this.trustDomain = trustDomain;
+        this.x509Authorities = ConcurrentHashMap.newKeySet();
+        this.x509Authorities.addAll(x509Authorities);
+    }
+
+    /**
+     * Loads a X.509 bundle from a file on disk.
+     *
+     * @param trustDomain a {@link TrustDomain} to associate to the bundle
+     * @param bundlePath  a path to the file that has the X.509 authorities
+     * @return an instance of {@link X509Bundle} with the X.509 authorities
+     * associated to the trust domain.
+     *
+     * @throws IOException in case of failure accessing the given bundle path
+     * @throws CertificateException if the bundle cannot be parsed
+     */
+    public static X509Bundle load(@NonNull final TrustDomain trustDomain, @NonNull final Path bundlePath) throws IOException, CertificateException {
+        byte[] bundleBytes;
+        try {
+            bundleBytes = Files.readAllBytes(bundlePath);
+        } catch (NoSuchFileException e) {
+            throw new IOException("Unable to load X.509 bundle file");
+        }
+        val x509Certificates = CertificateUtils.generateCertificates(bundleBytes);
+        val x509CertificateSet = new HashSet<>(x509Certificates);
+        return new X509Bundle(trustDomain, x509CertificateSet);
+    }
+
+    /**
+     * Parses a X095 bundle from an array of bytes.
+     *
+     * @param trustDomain a {@link TrustDomain} to associate to the X.509 bundle
+     * @param bundleBytes an array of bytes that represents the X.509 authorities
+     *
+     * @return an instance of {@link X509Bundle} with the X.509 authorities
+     * associated to the given trust domain
+     *
+     * @throws CertificateException if the bundle cannot be parsed
+     */
+    public static X509Bundle parse(@NonNull final TrustDomain trustDomain, @NonNull final byte[] bundleBytes) throws CertificateException {
+        val x509Certificates = CertificateUtils.generateCertificates(bundleBytes);
+        val x509CertificateSet = new HashSet<>(x509Certificates);
+        return new X509Bundle(trustDomain, x509CertificateSet);
+    }
+
+    /**
+     * Returns the X.509 bundle associated to the trust domain.
+     *
+     * @param trustDomain an instance of a {@link TrustDomain}
+     * @return the {@link X509Bundle} associated to the given trust domain
+     *
+     * @throws BundleNotFoundException if no X.509 bundle can be found for the given trust domain
+     */
+    @Override
+    public X509Bundle getBundleForTrustDomain(TrustDomain trustDomain) throws BundleNotFoundException {
+        if (this.trustDomain.equals(trustDomain)) {
+            return this;
+        }
+        throw new BundleNotFoundException(String.format("No X509 bundle found for trust domain %s", trustDomain));
+    }
+
+    /**
+     * Returns the X.509 x509Authorities in the bundle.
+     */
+    public Set<X509Certificate> getX509Authorities() {
+        return new HashSet<>(x509Authorities);
```

Comment:
Might be safer to return a read-only version of the field to prevent modification on the caller side using `Collections.unmodifiableSet()`

Reply:
I totally agree.